File: scripts/import-clientes-csv.js
/**
 * Script para importar clientes do CSV para o Supabase
 * Execute: node scripts/import-clientes-csv.js
 */

import { createClient } from '@supabase/supabase-js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configuração do Supabase
const supabaseUrl = process.env.VITE_SUPABASE_URL || '';
const supabaseKey = process.env.VITE_SUPABASE_ANON_KEY || '';

if (!supabaseUrl || !supabaseKey) {
  console.error('❌ Erro: Configure as variáveis de ambiente VITE_SUPABASE_URL e VITE_SUPABASE_ANON_KEY');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

// Função para parsear CSV
function parseCSV(content) {
  const lines = content.split('\n').filter(line => line.trim());
  if (lines.length < 2) return [];
  
  // Headers exatos do CSV (com maiúsculas e espaços)
  const headerLine = lines[0];
  const headers = headerLine.split(';').map(h => h.trim());
  const data = [];
  
  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(';');
    const row = {};
    
    headers.forEach((header, index) => {
      row[header] = values[index]?.trim() || null;
    });
    
    data.push(row);
  }
  
  return data;
}

// Função para mapear dados do CSV para o formato do banco
function mapCSVToCliente(row) {
  // Mapear colunas do CSV para o banco (usando nomes exatos do CSV)
  const codigo = row['Codigo'] || null;
  const nome = row['N Fantasia'] || row['Nome'] || null; // N Fantasia é o nome do cliente
  const nomeCompleto = row['Nome'] || null; // Nome completo da empresa
  const endereco = row['Endereco'] || null;
  const complemento = row['Complemento'] || null;
  const bairro = row['Bairro'] || null;
  const cidade = row['Municipio'] || null;
  const uf = row['Estado'] || null;
  const cep = row['CEP'] || null;
  const cnpjCpf = row['CNPJ/CPF'] || null;
  const insEstadual = row['Ins. Estad.'] || null;
  const bloqueado = row['Bloqueado'] || null;
  const codigoMunicipio = row['Cd.Municipio'] || null;
  const pais = row['Pais'] || null;
  const vendedor = row['Vendedores'] || null;
  const loja = row['Loja'] || null;
  
  // Extrair rede do nome (ex: "ATAC BARROS REIS" -> rede pode ser "ATACADAO")
  let rede = null;
  if (nome) {
    if (nome.includes('ATAC') || nome.includes('ATACADAO')) rede = 'ATACADAO';
    else if (nome.includes('ASSAI')) rede = 'ASSAI';
    else if (nome.includes('ATAKAREJO')) rede = 'ATAKAREJO';
    else if (nome.includes('GBARBOSA') || nome.includes('G BARBOSA')) rede = 'G BARBOSA';
    else if (nome.includes('RMIX') || nome.includes('REDEMIX')) rede = 'REDE MIX';
    else if (nome.includes('HIPERIDEAL') || nome.includes('SERRANA')) rede = 'HIPERIDEAL';
    else if (nome.includes('MATEUS')) rede = 'MATEUS';
    else if (nome.includes('PERINI')) rede = 'PERINI';
    else if (nome.includes('CBD')) rede = 'COMPANHIA BRASILEIRA DE DISTRIBUICAO';
    else if (nome.includes('SAMS')) rede = 'SAMS';
    else if (nome.includes('CDP')) rede = 'COSTA DO PLAZA';
  }
  
  // Converter latitude e longitude
  let latitude = null;
  let longitude = null;
  
  if (row['Latitude']) {
    const lat = parseFloat(String(row['Latitude']).replace(/,/g, '.'));
    if (!isNaN(lat)) latitude = lat;
  }
  
  if (row['Longitude']) {
    const lng = parseFloat(String(row['Longitude']).replace(/,/g, '.'));
    if (!isNaN(lng)) longitude = lng;
  }
  
  if (!nome) return null; // Pular se não tiver nome
  
  return {
    codigo,
    nome,
    endereco,
    complemento,
    bairro,
    cidade,
    uf,
    cep,
    cnpj_cpf: cnpjCpf,
    ins_estadual: insEstadual,
    bloqueado,
    codigo_municipio: codigoMunicipio,
    pais,
    loja,
    rede,
    vendedor,
    latitude,
    longitude,
  };
}

// Função principal
async function importarClientes() {
  try {
    console.log('🔄 Iniciando importação de clientes...');
    
    // Ler arquivo CSV
    const csvPath = path.join(__dirname, '../ARQUIVOS/enderecos-clientesCSV.csv');
    const csvContent = fs.readFileSync(csvPath, 'utf-8');
    
    console.log('📄 Arquivo CSV lido com sucesso');
    
    // Parsear CSV
    const csvData = parseCSV(csvContent);
    console.log(`📊 ${csvData.length} linhas encontradas no CSV`);
    
    // Mapear para formato do banco
    const clientes = csvData
      .map(mapCSVToCliente)
      .filter(c => c !== null);
    
    console.log(`✅ ${clientes.length} clientes válidos para importar`);
    
    // Inserir em lotes de 50
    const batchSize = 50;
    let inserted = 0;
    let updated = 0;
    let errors = 0;
    
    for (let i = 0; i < clientes.length; i += batchSize) {
      const batch = clientes.slice(i, i + batchSize);
      
      try {
        // Usar upsert para atualizar se já existir (baseado no codigo que é UNIQUE)
        const { data, error } = await supabase
          .from('clientes')
          .upsert(batch, {
            onConflict: 'codigo',
            ignoreDuplicates: false
          });
        
        if (error) {
          console.error(`❌ Erro ao inserir lote ${Math.floor(i / batchSize) + 1}:`, error);
          errors += batch.length;
        } else {
          inserted += batch.length;
          console.log(`✅ Lote ${Math.floor(i / batchSize) + 1} processado: ${batch.length} clientes`);
        }
      } catch (err) {
        console.error(`❌ Erro ao processar lote ${Math.floor(i / batchSize) + 1}:`, err);
        errors += batch.length;
      }
    }
    
    console.log('\n📊 Resumo da importação:');
    console.log(`✅ ${inserted} clientes processados`);
    console.log(`❌ ${errors} erros`);
    console.log('🎉 Importação concluída!');
    
  } catch (error) {
    console.error('❌ Erro na importação:', error);
    process.exit(1);
  }
}

// Executar importação
importarClientes();

File: scripts/import-clientes-csv.py
#!/usr/bin/env python3
"""
Script para importar clientes do CSV para o Supabase via SQL direto
"""

import csv
import sys
import os
from pathlib import Path

# Caminho do arquivo CSV
csv_path = Path(__file__).parent.parent / 'ARQUIVOS' / 'enderecos-clientesCSV.csv'

def parse_csv():
    """Lê e parseia o arquivo CSV"""
    clientes = []
    
    with open(csv_path, 'r', encoding='utf-8') as f:
        # Ler CSV com delimitador ponto e vírgula
        reader = csv.DictReader(f, delimiter=';')
        
        for row in reader:
            # Extrair dados
            codigo = row.get('Codigo', '').strip() or None
            nome_fantasia = row.get('N Fantasia', '').strip() or row.get('Nome', '').strip() or None
            
            if not nome_fantasia:
                continue  # Pular se não tiver nome
            
            # Mapear colunas
            cliente = {
                'codigo': codigo,
                'nome': nome_fantasia,
                'endereco': row.get('Endereco', '').strip() or None,
                'complemento': row.get('Complemento', '').strip() or None,
                'bairro': row.get('Bairro', '').strip() or None,
                'cidade': row.get('Municipio', '').strip() or None,
                'uf': row.get('Estado', '').strip() or None,
                'cep': row.get('CEP', '').strip() or None,
                'pais': row.get('Pais', '').strip() or None,
                'cnpj_cpf': row.get('CNPJ/CPF', '').strip() or None,
                'ins_estadual': row.get('Ins. Estad.', '').strip() or None,
                'bloqueado': row.get('Bloqueado', '').strip() or None,
                'codigo_municipio': row.get('Cd.Municipio', '').strip() or None,
                'loja': row.get('Loja', '').strip() or None,
                'vendedor': row.get('Vendedores', '').strip() or None,
            }
            
            # Converter latitude e longitude
            lat_str = row.get('Latitude', '').strip()
            lng_str = row.get('Longitude', '').strip()
            
            try:
                if lat_str:
                    lat = float(lat_str.replace(',', '.'))
                    cliente['latitude'] = lat
                else:
                    cliente['latitude'] = None
            except (ValueError, AttributeError):
                cliente['latitude'] = None
            
            try:
                if lng_str:
                    lng = float(lng_str.replace(',', '.'))
                    cliente['longitude'] = lng
                else:
                    cliente['longitude'] = None
            except (ValueError, AttributeError):
                cliente['longitude'] = None
            
            # Inferir rede do nome
            nome_upper = nome_fantasia.upper()
            rede = None
            if 'ATAC' in nome_upper or 'ATACADAO' in nome_upper:
                rede = 'ATACADAO'
            elif 'ASSAI' in nome_upper:
                rede = 'ASSAI'
            elif 'ATAKAREJO' in nome_upper:
                rede = 'ATAKAREJO'
            elif 'GBARBOSA' in nome_upper or 'G BARBOSA' in nome_upper:
                rede = 'G BARBOSA'
            elif 'RMIX' in nome_upper or 'REDEMIX' in nome_upper:
                rede = 'REDE MIX'
            elif 'HIPERIDEAL' in nome_upper or 'SERRANA' in nome_upper:
                rede = 'HIPERIDEAL'
            elif 'MATEUS' in nome_upper:
                rede = 'MATEUS'
            elif 'PERINI' in nome_upper:
                rede = 'PERINI'
            elif 'CBD' in nome_upper:
                rede = 'COMPANHIA BRASILEIRA DE DISTRIBUICAO'
            elif 'SAMS' in nome_upper:
                rede = 'SAMS'
            elif 'CDP' in nome_upper:
                rede = 'COSTA DO PLAZA'
            
            cliente['rede'] = rede
            clientes.append(cliente)
    
    return clientes

def generate_sql(clientes):
    """Gera SQL para inserir/atualizar clientes"""
    sql_statements = []
    
    for cliente in clientes:
        # Escapar aspas simples nos valores
        def escape(value):
            if value is None:
                return 'NULL'
            if isinstance(value, (int, float)):
                return str(value)
            return f"'{str(value).replace(\"'\", \"''\")}'"
        
        # Construir SQL INSERT com ON CONFLICT
        sql = f"""
INSERT INTO public.clientes (
    codigo, nome, endereco, complemento, bairro, cidade, uf, cep, pais,
    cnpj_cpf, ins_estadual, bloqueado, codigo_municipio, loja, vendedor,
    rede, latitude, longitude
) VALUES (
    {escape(cliente['codigo'])},
    {escape(cliente['nome'])},
    {escape(cliente['endereco'])},
    {escape(cliente['complemento'])},
    {escape(cliente['bairro'])},
    {escape(cliente['cidade'])},
    {escape(cliente['uf'])},
    {escape(cliente['cep'])},
    {escape(cliente['pais'])},
    {escape(cliente['cnpj_cpf'])},
    {escape(cliente['ins_estadual'])},
    {escape(cliente['bloqueado'])},
    {escape(cliente['codigo_municipio'])},
    {escape(cliente['loja'])},
    {escape(cliente['vendedor'])},
    {escape(cliente['rede'])},
    {escape(cliente['latitude'])},
    {escape(cliente['longitude'])}
)
ON CONFLICT (codigo) DO UPDATE SET
    nome = EXCLUDED.nome,
    endereco = EXCLUDED.endereco,
    complemento = EXCLUDED.complemento,
    bairro = EXCLUDED.bairro,
    cidade = EXCLUDED.cidade,
    uf = EXCLUDED.uf,
    cep = EXCLUDED.cep,
    pais = EXCLUDED.pais,
    cnpj_cpf = EXCLUDED.cnpj_cpf,
    ins_estadual = EXCLUDED.ins_estadual,
    bloqueado = EXCLUDED.bloqueado,
    codigo_municipio = EXCLUDED.codigo_municipio,
    loja = EXCLUDED.loja,
    vendedor = EXCLUDED.vendedor,
    rede = EXCLUDED.rede,
    latitude = EXCLUDED.latitude,
    longitude = EXCLUDED.longitude;
"""
        sql_statements.append(sql.strip())
    
    return sql_statements

if __name__ == '__main__':
    print('🔄 Processando CSV...')
    clientes = parse_csv()
    print(f'✅ {len(clientes)} clientes encontrados no CSV')
    
    print('📝 Gerando SQL...')
    sql_statements = generate_sql(clientes)
    
    # Salvar SQL em arquivo para execução
    sql_file = Path(__file__).parent.parent / 'scripts' / 'import_clientes.sql'
    with open(sql_file, 'w', encoding='utf-8') as f:
        f.write('-- Script gerado automaticamente para importar clientes do CSV\n')
        f.write('-- Execute este script no Supabase SQL Editor\n\n')
        for sql in sql_statements:
            f.write(sql + '\n\n')
    
    print(f'✅ SQL gerado em: {sql_file}')
    print(f'📊 Total de {len(sql_statements)} comandos SQL gerados')
    print('\n💡 Execute o arquivo SQL no Supabase SQL Editor ou use o MCP do Supabase')

